### PR TITLE
FA-84 Added the ability to get Token from database.

### PR DESCRIPTION
**Pull request by HamdanSCode**
also addressed some things that were meant to also be handled in this ticket like removing delays.

### CODE REVIEW

File: app/src/main/java/com/hamdan/forzenbook/data/database/LoginDao.kt [hunk 1]
```diff
@@ -17,4 +17,13 @@ interface LoginDao {
         """
     )
     suspend fun getToken(): LoginEntity?
+
+    // TODO When implementing further points for the token usage
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/database/LoginDao.kt]: Where ticket number?

File: app/src/main/java/com/hamdan/forzenbook/data/database/LoginEntity.kt [hunk 2]
```diff
@@ -5,9 +5,11 @@ import androidx.room.Entity
 import androidx.room.PrimaryKey
 import com.hamdan.forzenbook.data.database.LoginEntity.Companion.TABLE_NAME
 
+// TODO when creating homepage login with token, may need to remove the email portion from this as only a token may be needed
 @Entity(tableName = TABLE_NAME)
 data class LoginEntity(
-    @ColumnInfo(name = TOKEN) val token: String?
+    @ColumnInfo(name = TOKEN) val token: String?,
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/database/LoginEntity.kt]: You don't need to make these nullable in your entity

File: app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
         service.requestValidation(email).isSuccessful
     }
 
-    override suspend fun getToken(email: String, code: String): LoginResponse {
+    override suspend fun getToken(): LoginData? {
+        // we catch the exception up the chain if there is an issue
+        return dao.getToken()?.toLoginData()
+    }
+
+    override suspend fun getToken(email: String, code: String): LoginData {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt]: The repo should be deciding which data method to retrieve from. I think it's best to combine these methods and let the repo choose to grab from db or go to the network. Either that or separate their naming?

File: app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromDatabaseUseCaseImpl.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.hamdan.forzenbook.domain.usecase.login
+
+import android.util.Log
+import com.hamdan.forzenbook.data.repository.LoginRepository
+
+class LoginGetTokenFromDatabaseUseCaseImpl(
+    val repository: LoginRepository
+) : LoginGetTokenFromDatabaseUseCase {
+    override suspend operator fun invoke() {
+        val token = repository.getToken()?.token ?: throw Exception("null token")
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromDatabaseUseCaseImpl.kt]: Make your own exception for this like in the other spots

File: app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.hamdan.forzenbook.domain.usecase.login
+
+import android.util.Log
+import com.hamdan.forzenbook.data.repository.LoginRepository
+
+class LoginGetTokenFromNetworkUseCaseImpl(
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt]: I don't like the naming here, similar to the repo methods naming. It doesn't really describe what's going on. This is a "LoginUseCase", the token is an implementation detail that doesn't need to be part of naming. the Database use case is "GetStoredCredentials", again with the token not needing to be part of the naming. 

Implementation details can be hidden away

File: app/src/main/java/com/hamdan/forzenbook/data/database/LoginDao.kt (hunk 1 above)
Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/data/database/LoginDao.kt]: A ticket for it doesn't exist yet, so I guess the ToDo can just be removed.

File: app/src/main/java/com/hamdan/forzenbook/data/database/LoginEntity.kt (hunk 2 above)
Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/data/database/LoginEntity.kt]: My mistake, I was thinking that if we could possibly get null from the db we should probably give it nullables as well, but thats just flawed thinking.

File: app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt (hunk 3 above)
Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt]: I think seperating their naming is the more viable option, it feels really pointless to purposely send down an empty email and code just so an if can see it so it does a different action.

File: app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromDatabaseUseCaseImpl.kt (hunk 4 above)
Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromDatabaseUseCaseImpl.kt]: created, makes sense to have the specific exception will make handling the error easier down the line when I actually implement sending to the homepage

File: app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt (hunk 5 above)
Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt]: Should I then put all the Login-based Use cases under one Use case? I thought that was against the point of the Use cases. giving explicitly what's needed. The naming definitely feels wrong but how do I differentiate them then?

Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt]: ok actually I kind of see what your saying on re-reading the comment

Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/domain/usecase/login/LoginGetTokenFromNetworkUseCaseImpl.kt]: renamed the use cases following those ideas, though I may be interpreting it wrong.

File: app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepository.kt [hunk 6]
```diff
@@ -1,9 +1,9 @@
 package com.hamdan.forzenbook.data.repository
 
-import com.hamdan.forzenbook.data.network.LoginResponse
-
 interface LoginRepository {
     suspend fun requestValidation(email: String)
 
-    suspend fun getToken(email: String, code: String): LoginResponse?
+    suspend fun getTokenFromNetwork(email: String, code: String): LoginData?
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepository.kt]: The repo should decide in one method where to get the token from, either network or database, like the other examples we have done

Comment (HamdanSCode) [reply app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepository.kt]: combined methods into one essentialy

File: app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt [hunk 7]
```diff
@@ -1,22 +1,38 @@
...
-    // TODO implement the dao FA-84
     private val dao: LoginDao,
     private val service: LoginService
 ) : LoginRepository {
     override suspend fun requestValidation(email: String) {
         service.requestValidation(email).isSuccessful
     }
 
-    override suspend fun getToken(email: String, code: String): LoginResponse {
+    override suspend fun getToken(email: String?, code: String?): LoginData? {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt]: Abstract away from front end layers

File: app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt [hunk 8]
```diff
@@ -1,22 +1,43 @@
...
-        } else throw Exception("Failed to get Token")
+            val body = response.body() ?: throw FailTokenRetrievalException("Failed to get Token")
+            body.token ?: throw NullTokenException("null token")
+            dao.insert(LoginEntity(body.token, email))
+            return User.LoggedIn
+        } else throw FailTokenRetrievalException("Failed to get Token")
     }
 }
+
+sealed interface User {
```
Comment (ForzenAcademy) [inline app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt]: Put these in own file

**Review by ForzenAcademy — APPROVED** [review]

**Pull request merged**